### PR TITLE
pull uses of context up as high as possible

### PR DESCRIPTION
**Pull request by mhseiden**
This PR should have no functional changes. It just pulls uses of `context` out of inner functions, and as close to public methods as possible.

### CODE REVIEW

File: heartbeat.go [hunk 1]
```diff
@@ -81,7 +81,7 @@ func (hc *heartbeat) heartbeatMain() error {
 			return err
 		}
 		if respd.Code == sessionExpiredCode {
-			err = hc.restful.FuncRenewSession(context.TODO(), hc.restful)
+			err = hc.restful.FuncRenewSession(context.TODO(), hc.restful, hc.restful.RequestTimeout)
```
Comment (donhcd) [inline heartbeat.go]: does this need to be in this patch?

File: restful.go [hunk 2]
```diff
@@ -110,20 +110,19 @@ func postRestfulQuery(
...
 
 	requestID := uuid.New().String()
-
 	data, err = sr.FuncPostQueryHelper(ctx, sr, params, headers, body, timeout, requestID)
 
-	// errors other than context timeout and cancel would be returned to upper layers
-	if err != context.Canceled && err != context.DeadlineExceeded {
-		return data, err
+	// If the context was canceled then we need to cancel the query as well. This is done
+	// in a separate goroutine, to that the caller's context deadline is respected.
```
Comment (donhcd) [inline restful.go]: so that?

**Comment by mhseiden** [comment]
Ah, right! I'll move those out into a separate PR.

**Comment by mhseiden** [comment]
Closing in favor of a PR + branch actually in this fork.

**Pull request closed without merging**